### ISSUE TITLE
Support for C-CANCEL requests for C-GET,MOVE,FIND

### ISSUE THREAD

**Issue by hkoestin** [fo-dicom/fo-dicom#748]
### Expected behavior
I would like to be able to send C-CANCEL requests easily.

### Actual behavior
There is the DicomCommandField for CCancel, but no class actually implementing that stuff.

### fo-dicom version and OS/platform
4.0.0

## Sketch of a small implementation
This class could sit in the package, to offer some nice convenience methods for the user.

```c#
/// <summary>
/// DICOM C-CANCEL request for a corresponding C-GET, C-FIND, or C-MOVE request.
/// </summary>
public sealed class DicomCCancelRequest : DicomPriorityRequest
{
    /// <inheritdoc />
    private DicomCCancelRequest(DicomRequest requestToCancel)
        : base(DicomCommandField.CCancelRequest, requestToCancel.SOPClassUID, DicomPriority.High)
    {
        Dataset = new DicomDataset();
        Dataset.AddOrUpdate(DicomTag.MessageIDBeingRespondedTo, requestToCancel.MessageID);
    }

    /// <inheritdoc />
    protected override void PostResponse(DicomService service, DicomResponse response)
    {
        // empty on purpose, as there is no response to this message
    }

    /// <summary>
    /// Creates a corresponding C-CANCEL request for the given C-MOVE request..
    /// </summary>
    public static DicomPriorityRequest Create(DicomCMoveRequest request)
    {
        return new DicomCCancelRequest(request);
    }

    /// <summary>
    /// Creates a corresponding C-CANCEL request for the given C-FIND request..
    /// </summary>
    public static DicomPriorityRequest Create(DicomCFindRequest request)
    {
        return new DicomCCancelRequest(request);
    }

    /// <summary>
    /// Creates a corresponding C-CANCEL request for the given C-GET request..
    /// </summary>
    public static DicomPriorityRequest Create(DicomCGetRequest request)
    {
        return new DicomCCancelRequest(request);
    }
}
```

**Comment by gofal** [comment]
please create a pull request to add this DicomCCancelRequest.

**Comment by gofal** [comment]
Have you already seen this issue: #624 
There was a discssion about c-cancel. But we figured out, that the handling of the c-cancel, namely interrupting a long running task, is hard to implement currently.
It would need some enhancement like #652 to enable sending such c-cancel on time. Currently you have to enqueue all requests you want to send in a list of DicomClient class and then by calling DicomClient.Send fire al the requests. I see no reasonable workflow how to handle the c-cancel. 

How would you use this request? Have you found a workflow when and how to send this c-cancel request?

**Comment by hkoestin** [comment]
Yes, I have seen that "issue" #624 and also #652.

On my application side, requests to my software can be canceled. So when the user e.g. starts to do a C-MOVE of a big study from the PACS to TARGET AE and figures out the operation should be canceled, it should be possible to cancel the C-MOVE request on the PACS. 

For canceling I use the concept of CancellationToken in .NET and simply attach an action to it, which, on a different DicomClient dispatches a C-CANCEL-C-MOVE-RQ to the PACS.

This would actually not be needed, as I could just simply ignore all further incoming C-STORE RQs from the PACS, which are related to the C-MOVE RQ I initiated. But to keep things tidy and clean, I just try to do my best and tell the PACS about the cancellation. In fact, this still depends on the implementation of the PACS then, if it supports the cancel after all or not. But most of the times (tested now with Dicoogle and Orthanc), they support it, and eventually, the C-CANCEL will get through in time to at least save some of the C-STORE RQs.

So for doing "best practice" C-CANCEL and telling the PACS it should stop sending the C-STOREs, this works fine and nice.

Here you can find a sketch for that
```c#

// Note
// m_requestDispatcher is an abstraction over a DicomClient, which creates a new Client
// for every dispatched request

public IList<DicomDataset> Retrieve(CancellationToken cancellationToken, string studyInstanceUid)
{
    var request = new DicomCMoveRequest(TargetAeTitle, studyInstanceUid);
    var cStoresFinished = new TaskCompletionSource<bool>();
    
    cancellationToken.Register(() =>
    {
        m_logger.Warn("Request was canceled. Stopping to wait for C-MOVE response and sending C-CANCEL-MOVE-RQ.");

        // According to the standard:
        // http://dicom.nema.org/Dicom/2013/output/chtml/part07/sect_9.3.html#sect_9.3.4.3
        // The C-MOVE procedures can be canceled at any time by the invoking DIMSE-service-user. 
        // This shall be accomplished by the invoking DIMSE-service-user issuing a C-CANCEL request primitive.
        m_requestDispatcher.Execute(CancellationToken.None, DicomCCancelRequest.Create(request));
    });

    // Note: there is a strict assumption in here
    // PACS must send the final "SUCCESS" response AFTER the last C-Store 
    // See standard: http://dicom.nema.org/medical/dicom/current/output/chtml/part04/sect_C.4.2.3.html

    request.OnResponseReceived += (moveRequest, response) =>
    {
        m_logger.Debug(response.ToString);

        if (response.Status == DicomStatus.Success) 
            cStoresFinished.SetResult(true);
        if (response.Status == DicomStatus.ProcessingFailure)
            cStoresFinished.SetResult(false);
    };

    try
    {
        // issue the actual C-MOVE request
        m_requestDispatcher.Execute(cancellationToken, request);

        // lets wait for the C-STOREs to come in and set the last result upon the C-MOVE response
        cStoresFinished.Task.Wait(cancellationToken);
    }
    catch (Exception exception)
    {
        m_logger.Error("C-MOVE request failed. Trying to transition task completion source to 'false' result state.", exception);
        cStoresFinished.TrySetResult(false);
    }

    if (cStoresFinished.Task.Result)
    {
        m_logger.Debug("C-MOVE successful and complete. Fetching received datasets from storage server.");
        return m_storageServer.FetchReceivedDatasetsForRequest(request);
    }

    return new List<DicomDataset>();
}
```

**Comment by aerik** [comment]
I think this is still relevant, as I don't see C-Cancel implemented... (?)  How about implement C-Cancel in DicomService base, keep a dictioanry of message IDs and cancellation token sources, then use that if a C-Cancel is received?  (I think you'd have to change teh signature of OnCFindRequest, etc, to include a cancellaion token, then if a C-Cancel is received, look up the cancellatino token source for that message and cancel.

**Comment by aerik** [comment]
Well that was naive of me.  It was a bit more complicated than I thought.  I ended up just looking up the request from _pending, the n using a volatile bool (added to DicomRequest) to set a cancelled status.

However, in my branch at least, the C-Cancel (and Aborts too, for that matter) are not picked up until AFTER the OnWhatever event handlers have finished (inside PerformDimse inside ProcessPDataTFAsync).  Even though these are async, they still prevent the next message from being picked up from the networkstream.

**Comment by gofal** [comment]
Great, thanks for digging into that.
I also did some research after your first post 2 days ago. The C-Cancel would have to be implemented in 2 places, the client side with the ability to send a c-cancel while a c-move, a c-get or a c-find is in progress , and the server side where such a c-cancel message is processed and notified to the application.

Maybe we could do a divide et impera and split this into two stories.
Because from what I assume, at least the client-side part would be more easy. 
I tried to figure out from the dicom-standard, if such a c-cancel message has to be sent via the same association as the referenced c-move, c-get or c-find request. because if the c-cancel can be sent via another tcp-connection then the implementation would be very easy. Else adding a Cancel-method to the DicomClient, where the DicomClient knows what the current request is, and sending a c-cancel request through the current network stream should also not be too hard.

And when it comes to the server-side, there is exactly that issue, that you also mentioned. There was a very short discussion many years ago #624 where I already pointed at the processing-issue, that Requests are received and processed in a serialized way and a cancel, that is sent via the same connection cannot interrupt the currently ongoing process.
Except the C-Cancel was sent on a separate association, then the DicomServer could check all the current DicomServices and inject a message there. But as said, I am not sure if that is fine according to DICOM Standard

So to sum up, I think the story, to implement c-cancel in the client, could be implemented now. And a separate story to implment c-cancel on server would have to wait until there is some rebuilding of the underlying network infrastructure.

**Comment by aerik** [comment]
According to ChatGPT, we are meant to infer that a cancel happens on the same Assocaition (I couldn't find it explicitly stated, but this make sense to me)

> Why Implicitly “Same Association”
> 
> Because:
> 
> The standard defines DIMSE messages (requests, responses, cancel requests) as being exchanged over a given Association. The Association is the transport link under which DIMSE messages flow.
> 
> The definitions of the “Request primitive”, “Cancel Request primitive”, etc., are defined in terms of that DIMSE protocol machine over that Association.
> 
> There is no provision for opening a separate association just for cancellation. Cancellation is treated simply as a later request primitive of the same operation context.
> 
> Thus the “on the same association” requirement is implicit — it’s basic to how DIMSE / PS3.7 works.

On my fork of v3, I used the DicomCCancelRequest class given above, and build this stuff out.  Would love feedback about whether I'm missing anything important in my approach. (spelled out in detail in #2023 and viewable in my fork repo)

**Comment by ben-fondu-envista** [comment]
I just ran into this issue as well. Tried implementing the DicomCCancelRequest locally, but when sending it from the response handler of the C-Find, everything just hangs and I get an association aborted exception.

Is there an easy way to do it without code changes in fo-dicom?

Specifically, we're doing C-FIND requests, and need to be able to limit how many responses we get. As far as I can tell, the only real way to do this is with a C-CANCEL-FIND.

**Comment by ben-fondu-envista** [comment]
Well, I did some more poking around. It's indeed not a simple issue to solve.

I think at the root of the issue, we'll have to separate this issue into SCU and SCP, I believe. Implementing it for the SCU should be much simpler.

In `DicomService.SendMessageAsync()`, the check for max async ops could be modified to something like this:
```csharp
var nextMsg = _msgQueue.Peek();
var nextMsgResponseMsgId = nextMsg.Command.GetSingleValueOrDefault(DicomTag.MessageIDBeingRespondedTo, (ushort?)null);

if (Association.MaxAsyncOpsInvoked > 0
    && _pending.Count(req => req.Type != DicomCommandField.CGetRequest && req.Type != DicomCommandField.NActionRequest && req.MessageID != nextMsgResponseMsgId)
    >= Association.MaxAsyncOpsInvoked)
{
    break;
}
```

This can no doubt be made a bit more robust and intelligent so that check is only done for actual cancel commands, it can handle other commands being in the queue, etc.
But as a proof of concept, it does work.

**Comment by gofal** [comment]
sure, splitting it into 2 stories will make sense.
